### PR TITLE
Feature/m0078 ddg 284 document new platform bsp connection options

### PR DESCRIPTION
**Pull request by frivieremicroej**

### CODE REVIEW

**Comment by apoudret** [comment]
it would be nice if empty platform projects created using `File > New > MicroEJ Platform Project…` did not have so many files (`bsp.properties`, `init-bsp.xml`, `deployInBSP.xml`, `deployToolBSPRun.microejTool`, `wbe_launchScriptFramework.jar`, `configuration.xml`).
Maybe we can merge all the required configuration to a single file (the `.platform` file?)

File: PlatformDeveloperGuide/platformCreation.rst [hunk 1]
```diff
@@ -0,0 +1,442 @@
...
+   :align: center
+
+   MicroEJ Platform Configuration Project Skeleton
+
+
+
+Groups / Modules Selection
+==========================
+
+Group
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:85]:
```suggestion
```

`Group` is probably a mistake, the sentence doesn't make sense with it.

File: PlatformDeveloperGuide/platformCreation.rst [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+click on the Show/Hide modules icon on the top-right of the frame. This
+will let you select/deselect on a per module basis. Note that individual
+module selection is not recommended.
+
+The description and contents of an item (group or module) are displayed
+beside the list on item selection.
+
+All the checked modules will be installed in the Platform.
+
+
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:101]: Maybe add a screenshot with the `Content` tab opened.

File: PlatformDeveloperGuide/platformCreation.rst [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+==============
+
+Principle
+---------
+
+Using a MicroEJ Platform, the user can compile a MicroEJ Application on that Platform. 
+The result of this compilation is a ``microejapp.o`` file.
+
+This file has to be linked with the MicroEJ Platform runtime file (``microejruntime.a``) 
+and a third-party C project, called the Board Support Package (BSP)
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:161]:
```suggestion
and a third-party C project, called the Board Support Package (BSP) ,
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 4]
```diff
@@ -0,0 +1,442 @@
...
+
+Using a MicroEJ Platform, the user can compile a MicroEJ Application on that Platform. 
+The result of this compilation is a ``microejapp.o`` file.
+
+This file has to be linked with the MicroEJ Platform runtime file (``microejruntime.a``) 
+and a third-party C project, called the Board Support Package (BSP) 
+to obtain the final binary file (MicroEJ Firmware).
+For more information, please consult the :ref:`MicroEJ build process overview <build_process_overview>`.
+
+BSP connection consists in defining the 4 folders where are located:
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:165]:
```suggestion
The BSP Connection can be configured by defining 4 folders where the following files are located:
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+- MicroEJ Platform runtime file (``microejruntime.a``, also available in the Platform ``lib`` folder).
+- MicroEJ Platform header files (``*.h``, also available in the Platform ``include`` folder).
+- BSP project :ref:`build script <bsp_connection_build_script>` file (``build.bat`` or ``build.sh``).
+
+Once the MicroEJ Application file (``microejapp.o``) is built, the files are then copied to these locations 
+and the ``build.bat`` or ``build.sh`` file is executed to produce the final executable file (``application.out``).
+
+.. note::
+
+   The final build stage to produce the executable file can be done outside MicroEJ SDK, and thus
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:177]:
```suggestion
   The final build stage to produce the executable file can be done outside of MicroEJ SDK, and thus 
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+   BSP connection configuration is only required in the following cases:
+
+   - Use MicroEJ SDK to produce the final executable file of a Mono-Sandbox Firmware (recommended).
+   - Use MicroEJ SDK to run a :ref:`MicroEJ Testsuite <testsuite>` on device.
+   - Build a Multi-Sandbox Firmware.
+
+.. _bsp_connection_cases:
+
+MicroEJ provides a flexible way to configure BSP connection to target any kind of projects, teams organizations and company build flows.
+For that, BSP connection can be configured either at MicroEJ Platform level or at MicroEJ Application level (or a mix of both).
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:189]:
```suggestion
MicroEJ provides a flexible way to configure the BSP connection to target any kind of projects, teams organizations and company build flows.
The BSP connection can be configured either at MicroEJ Platform level or at MicroEJ Application level (or a mix of both). 
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+   - Use MicroEJ SDK to produce the final executable file of a Mono-Sandbox Firmware (recommended).
+   - Use MicroEJ SDK to run a :ref:`MicroEJ Testsuite <testsuite>` on device.
+   - Build a Multi-Sandbox Firmware.
+
+.. _bsp_connection_cases:
+
+MicroEJ provides a flexible way to configure BSP connection to target any kind of projects, teams organizations and company build flows.
+For that, BSP connection can be configured either at MicroEJ Platform level or at MicroEJ Application level (or a mix of both). 
+
+The following list describes the 3 most common integration cases:
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:191]:
```suggestion
The 3 most common integration cases are:
```

File: PlatformDeveloperGuide/platformCreation.rst (hunk 4 above)
Comment (gjadi-mej) [reply PlatformDeveloperGuide/platformCreation.rst:165]: Isn't there 5 folders?

* microejapp.relative.dir
* microejlib.relative.dir
* microejinc.relative.dir
* microejscript.relative.dir
* root.dir

File: PlatformDeveloperGuide/platformCreation.rst [hunk 8]
```diff
@@ -0,0 +1,442 @@
...
+
+- Case 1: No BSP connection
+
+  The MicroEJ Platform does not know the BSP at all.
+  BSP connection can be configured when building the MicroEJ Application (options with absolute paths).
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:196]:
```suggestion
  BSP connection can be configured when building the MicroEJ Application (folders defined with absolute paths).
```

The name `options` isn't used before. I think that either we mention it ealier when we introduce the 4 (or 5) folders to define, or use the term `folder` instead.

File: PlatformDeveloperGuide/platformCreation.rst [hunk 9]
```diff
@@ -0,0 +1,442 @@
...
+
+  This case is recommended when:
+
+  - the MicroEJ Firmware is built outside MicroEJ SDK.
+  - the same MicroEJ Platform is intended to be reused on multiple BSP projects which do not share the same structure.
+
+- Case 2: Partial BSP connection
+  
+  The MicroEJ Platform knows how the BSP is structured.
+  BSP connection is configured when building the MicroEJ Platform options (options with relative paths within the BSP),
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:213]:
```suggestion
  BSP connection is configured when building the MicroEJ Platform folders (folders with relative paths within the BSP), 
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 10]
```diff
@@ -0,0 +1,442 @@
...
+  This case is recommended when:
+
+  - the MicroEJ Firmware is built outside MicroEJ SDK.
+  - the same MicroEJ Platform is intended to be reused on multiple BSP projects which do not share the same structure.
+
+- Case 2: Partial BSP connection
+  
+  The MicroEJ Platform knows how the BSP is structured.
+  BSP connection is configured when building the MicroEJ Platform options (options with relative paths within the BSP), 
+  and the BSP root location is configured when building the MicroEJ Application (option with absolute path).
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:214]:
```suggestion
  and the BSP root location is configured when building the MicroEJ Application (folder with absolute path).
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+  This case is recommended when:
+  
+  - the MicroEJ Platform is used to build one MicroEJ Application on top of one BSP. 
+  - the Application and BSP are slightly coupled, thus making a change in the BSP just require to rebuild the firmware.
+
+- Case 3: Full BSP connection
+  
+  The MicroEJ Platform includes its BSP.
+  BSP connection is configured when building MicroEJ Platform options (options with relative paths within the Platform and the BSP), 
+  No BSP path need to be configured when building the MicroEJ Application.
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:232]:
```suggestion
  BSP connection is configured when building MicroEJ Platform folders (folders with relative paths within the Platform and the BSP), 
  No BSP path needs to be configured when building the MicroEJ Application.
```

Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:232]:
```suggestion
  No BSP path needs to be configured when building the MicroEJ Application.
```

File: PlatformDeveloperGuide/platformCreation.rst (hunk 8 above)
Comment (gjadi-mej) [reply PlatformDeveloperGuide/platformCreation.rst:196]: Ok, so the options are described a bit later. The remark still stands: it's confusing to use the term `options` before it is defined. I think all the mentions to options in the overview should be removed. How to configure each option is described later.

File: PlatformDeveloperGuide/platformCreation.rst (hunk 10 above)
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:214]:
```suggestion
  BSP connection is configured when building the MicroEJ Platform, 
  and the BSP root location is configured when building the MicroEJ Application.
```

File: PlatformDeveloperGuide/platformCreation.rst (hunk 11 above)
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:232]:
```suggestion
  BSP connection is configured when building MicroEJ Platform options.
  No BSP path needs to be configured when building the MicroEJ Application.
```

File: PlatformDeveloperGuide/platformCreation.rst [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+The BSP build script file is responsible to invoke the third-party C toolchain (compiler and linker)
+to produce the final executable file (``application.out``).
+
+The build script must implement the following specification:
+
+- On Windows operating system, it is a Windows batch file named ``build.bat``.
+- On Mac OS X or Linux operating systems, it is a shell script named ``build.sh``, with execution permission enabled.
+- On build error, the script must end with a non zero exit code.
+- On success, the executable must be copied to the file ``application.out`` in the directory from
+  where the script has been executed. Then the script must end with zero exit code.
```
Comment (gjadi-mej) [inline PlatformDeveloperGuide/platformCreation.rst:389]:
```suggestion
- On success

  - The executable must be copied to the file ``application.out`` in the directory from
  where the script has been executed.
  - The script must end with zero exit code.
```

File: PlatformDeveloperGuide/platformCreation.rst (hunk 4 above)
Comment (frivieremicroej) [reply PlatformDeveloperGuide/platformCreation.rst:164]: no there are 4 folders, maybe defined with these 5 options (4 relative + 1 absolute) or by 4 absolute application options

File: PlatformDeveloperGuide/platformCreation.rst (hunk 8 above)
Comment (frivieremicroej) [reply PlatformDeveloperGuide/platformCreation.rst:195]: I will reuse terms of the schematic, without speaking about "options"

File: PlatformDeveloperGuide/platformCreation.rst (hunk 10 above)
Comment (frivieremicroej) [reply PlatformDeveloperGuide/platformCreation.rst:213]: I will reuse terms of the schematic, without speaking about "options"

File: PlatformDeveloperGuide/platformCreation.rst (hunk 11 above)
Comment (frivieremicroej) [reply PlatformDeveloperGuide/platformCreation.rst:231]: I will reuse terms of the schematic, without speaking about "options"

File: PlatformDeveloperGuide/platformCreation.rst (hunk 2 above)
Comment (frivieremicroej) [reply PlatformDeveloperGuide/platformCreation.rst:106]: done

**Pull request merged**